File: src/main/java/com/janboucek/crawler/iohandling/Logger.java
package com.janboucek.crawler.iohandling;

import com.janboucek.crawler.simulation.FitnessResult;
import com.janboucek.crawler.testsettings.TestSettings;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;

/**
 * Created by colander on 2/5/17.
 * Logs debug messages and the results of an evolution run.
 */
public class Logger {
    public final static String RESULTS_DIRECTORY = "results/";
    private final String runDir;

    private BufferedWriter logWriter;

    public Logger() {
        File resultsDir = new File(RESULTS_DIRECTORY);
        if (!resultsDir.exists()) IOHandler.createDirectory(resultsDir.getAbsolutePath());
        runDir = RESULTS_DIRECTORY + System.currentTimeMillis();
        IOHandler.createDirectory(runDir);
        File logFile = new File(runDir + "/evolution.log");
        try {
            logWriter = new BufferedWriter(new FileWriter(logFile));
        } catch (IOException e) {
            e.printStackTrace();
        }
        IOHandler.writeFile(runDir + "/config.cfg", TestSettings.serialize());
    }

    public void logGeneration(ArrayList<FitnessResult> results, int generationNo) {
        //make a copy which we can reverse without changing the original list
        ArrayList<FitnessResult> modifiableResults = new ArrayList<>();
        modifiableResults.addAll(results);

        Collections.reverse(modifiableResults); //reverse, so that the first genotype has the highest fitness
        StringBuilder sb = new StringBuilder();
        for (FitnessResult result : modifiableResults) {
            sb.append(result.result).append("\n").append(result.genotype.serialize()).append("\n");
        }
        IOHandler.writeFile(runDir + "/" + String.format("%04d", generationNo) + ".gen", sb.toString());
    }

    public void log(String message) {
        String[] split = message.split("\n");
        for (String s : split) {
            System.out.println(System.currentTimeMillis() + "|" + s);
            try {
                logWriter.write(System.currentTimeMillis() + "|" + s);
                logWriter.newLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

    }

    public void flush() {
        try {
            logWriter.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/java/com/janboucek/crawler/iohandling/Logger.kt
package com.janboucek.crawler.iohandling

import com.janboucek.crawler.iohandling.IOHandler.createDirectory
import com.janboucek.crawler.iohandling.IOHandler.writeFile
import com.janboucek.crawler.simulation.FitnessResult
import com.janboucek.crawler.testsettings.TestSettings
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException

/**
 * Created by colander on 2/5/17.
 * Logs debug messages and the results of an evolution run.
 */
class Logger(val runId: Long) {
    companion object {
        const val RESULTS_DIRECTORY = "results/"
    }

    private val runDir: String
    private val logWriter: BufferedWriter

    init {
        val resultsDir = File(RESULTS_DIRECTORY)
        if (!resultsDir.exists()) resultsDir.mkdirs()
        if (!resultsDir.isDirectory) throw IllegalStateException("failed to create results folder")
        runDir = "$RESULTS_DIRECTORY${System.currentTimeMillis() / 1000}$runId"
        createDirectory(runDir)
        val logFile = File("$runDir/evolution.log")
        logWriter = BufferedWriter(FileWriter(logFile))
        writeFile("$runDir/config.cfg", TestSettings.serialize())
    }

    fun logGeneration(results: List<FitnessResult>, generationNo: Int) {
        val sb = StringBuilder()
        for (result in results.reversed()) {
            sb.append(result.result).append("\n").append(result.genotype.serialize()).append("\n")
        }
        writeFile(runDir + "/" + String.format("%04d", generationNo) + ".gen", sb.toString())
    }

    fun log(message: String) {
        val split = message.split("\n")
        for (s in split) {
            println(System.currentTimeMillis().toString() + "|" + s)
            try {
                logWriter.write(System.currentTimeMillis().toString() + "|" + s)
                logWriter.newLine()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    fun flush() {
        try {
            logWriter.flush()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}
